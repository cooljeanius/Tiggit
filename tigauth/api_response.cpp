#include "api_response.hpp"
#include <spread/misc/readjson.hpp>
#include <stdexcept>

using namespace TigAuth::API;

Response::Response(const std::string &str)
{
  clear();
  if(str != "")
    decode(str);
}

void Response::clear()
{
  isValid = false;
  isError = false;
  isAuthFail = false;
  hasUserInfo = false;
  newKey.clear();
  message.clear();
}

void Response::decode(const std::string &input)
{
  clear();

  // Get value as JSON
  Json::Value val = ReadJson::strToJson(input);

  // Get response type
  std::string str = val["type"].asString();

  if(str == "error") isError = true;
  else if(str == "authfail")
    {
      isError = true;
      isAuthFail = true;
    }
  else if(str == "message") {}
  else
    throw std::runtime_error("Unexpected server response: " + ((str=="")?input:str));

  // Decode other parts of the message

  if(val.isMember("newkey"))
    newKey = val["newkey"].asString();

  if(val.isMember("message"))
    message = val["message"].asString();

  if(val.isMember("generated"))
    generated = val["generated"].asString();

  if(val.isMember("userinfo"))
    {
      hasUserInfo = true;

      Json::Value usr = val["userinfo"];

      userid = usr["userid"].asString();
      usernick = usr["nickname"].asString();
      userauth = usr["authname"].asString();
      useritems = usr["items"].asString();
    }

  // If we get here, we have successfully decoded the response.
  isValid = true;
}
