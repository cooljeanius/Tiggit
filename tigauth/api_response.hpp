#ifndef __TIGAUTH_API_RESPONSE_HPP_
#define __TIGAUTH_API_RESPONSE_HPP_

#include <string>

namespace TigAuth
{
  namespace API
  {
    struct Response
    {
      bool isValid, isError, isAuthFail, hasUserInfo;
      std::string newKey, message, userid, usernick, userauth, useritems, generated;

      Response(const std::string &str="");
      void decode(const std::string &input);
      void clear();
    };
  }
}

#endif
