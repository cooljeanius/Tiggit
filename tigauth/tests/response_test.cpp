#include "api_response.hpp"
#include <iostream>

using namespace std;

void test(const std::string &str)
{
  cout << "\nTESTING: '" << str << "'\n";
  TigAuth::API::Response res;
  try { res.decode(str); }
  catch(exception &e)
    {
      cout << "CAUGHT: " << e.what() << endl;
      return;
    }
  if(!res.isValid)
    {
      cout << "Invalid response\n";
      return;
    }

  cout << " isError=" << res.isError
       << " isAuthFail=" << res.isAuthFail
       << " hasUserInfo=" << res.hasUserInfo << endl
       << " newKey=" << res.newKey
       << " message=" << res.message << endl
       << " userid=" << res.userid
       << " usernick=" << res.usernick
       << " userauth=" << res.userauth
       << " useritems=" << res.useritems << endl
       << " generated=" << res.generated << endl;
}

int main()
{
  test("");
  test("{}");
  test("{\"type\":\"fishcat\"}");
  test("{\"type\":\"message\"}");
  test("{\"type\":\"error\"}");
  test("{\"type\":\"authfail\",\"generated\":\"now\"}");

  test("{\"type\":\"message\",\"newkey\":\"NEW KEY!\"}");
  test("{\"type\":\"message\",\"message\":\"This is a message\"}");
  test("{\"type\":\"error\",\"message\":\"This is an error message\"}");
  test("{\"type\":\"message\",\"userinfo\":{\"userid\":\"1234\",\"nickname\":\"My Name is Nick\"}}");

  test("{\"type\":\"message\",\"message\":\"MESSAGE\",\"userinfo\":{\"userid\":\"1234\",\"nickname\":\"My Name is Nick\",\"authname\":\"Home\",\"items\":\"item1+item2+another_item+anitemIreallylike\"},\"newkey\":\"NEWKEY\",\"generated\":\"2013-01-22 15:47:12 UTC\"}");
  return 0;
}
