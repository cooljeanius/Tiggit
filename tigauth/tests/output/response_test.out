
TESTING: ''
CAUGHT: * Line 1, Column 1
  Syntax error: value, object or array expected.


TESTING: '{}'
CAUGHT: Unexpected server response: {}

TESTING: '{"type":"fishcat"}'
CAUGHT: Unexpected server response: fishcat

TESTING: '{"type":"message"}'
 isError=0 isAuthFail=0 hasUserInfo=0
 newKey= message=
 userid= usernick= userauth= useritems=
 generated=

TESTING: '{"type":"error"}'
 isError=1 isAuthFail=0 hasUserInfo=0
 newKey= message=
 userid= usernick= userauth= useritems=
 generated=

TESTING: '{"type":"authfail","generated":"now"}'
 isError=1 isAuthFail=1 hasUserInfo=0
 newKey= message=
 userid= usernick= userauth= useritems=
 generated=now

TESTING: '{"type":"message","newkey":"NEW KEY!"}'
 isError=0 isAuthFail=0 hasUserInfo=0
 newKey=NEW KEY! message=
 userid= usernick= userauth= useritems=
 generated=

TESTING: '{"type":"message","message":"This is a message"}'
 isError=0 isAuthFail=0 hasUserInfo=0
 newKey= message=This is a message
 userid= usernick= userauth= useritems=
 generated=

TESTING: '{"type":"error","message":"This is an error message"}'
 isError=1 isAuthFail=0 hasUserInfo=0
 newKey= message=This is an error message
 userid= usernick= userauth= useritems=
 generated=

TESTING: '{"type":"message","userinfo":{"userid":"1234","nickname":"My Name is Nick"}}'
 isError=0 isAuthFail=0 hasUserInfo=1
 newKey= message=
 userid=1234 usernick=My Name is Nick userauth= useritems=
 generated=

TESTING: '{"type":"message","message":"MESSAGE","userinfo":{"userid":"1234","nickname":"My Name is Nick","authname":"Home","items":"item1+item2+another_item+anitemIreallylike"},"newkey":"NEWKEY","generated":"2013-01-22 15:47:12 UTC"}'
 isError=0 isAuthFail=0 hasUserInfo=1
 newKey=NEWKEY message=MESSAGE
 userid=1234 usernick=My Name is Nick userauth=Home useritems=item1+item2+another_item+anitemIreallylike
 generated=2013-01-22 15:47:12 UTC
